mono/metadata/w32file-win32.c

```diff
@@ -81,23 +81,61 @@ mono_w32file_delete (const gunichar2 *name)
 	return res;
 }
 
+static void
+abort_w32_synchronous_io (HANDLE hThread)
+{
+	CancelSynchronousIo (hThread);
+}
+
 gboolean
-mono_w32file_read(gpointer handle, gpointer buffer, guint32 numbytes, guint32 *bytesread)
+mono_w32file_read (gpointer handle, gpointer buffer, guint32 numbytes, guint32 *bytesread)
 {
-	gboolean res;
+	gboolean res = FALSE;
+	gboolean interrupted;
+	HANDLE duplicate;
+
+	if (!DuplicateHandle (GetCurrentProcess (), GetCurrentThread (), GetCurrentProcess (), &duplicate, THREAD_TERMINATE, FALSE, 0))
+		return FALSE;
+
+	mono_thread_info_install_interrupt (abort_w32_synchronous_io, duplicate, &interrupted);
+	if (interrupted)
+		goto cleanup;
+
 	MONO_ENTER_GC_SAFE;
 	res = ReadFile (handle, buffer, numbytes, bytesread, NULL);
 	MONO_EXIT_GC_SAFE;
+
+	mono_thread_info_uninstall_interrupt (&interrupted);
+
+cleanup:
+	CloseHandle (duplicate);
+
 	return res;
 }
 
 gboolean
 mono_w32file_write (gpointer handle, gconstpointer buffer, guint32 numbytes, guint32 *byteswritten)
 {
-	gboolean res;
+	gboolean res = FALSE;
+	gboolean interrupted;
+	HANDLE duplicate;
+
+	if (!DuplicateHandle (GetCurrentProcess (), GetCurrentThread (), GetCurrentProcess (), &duplicate, THREAD_TERMINATE, FALSE, 0))
+		return FALSE;
+
+	mono_thread_info_install_interrupt (abort_w32_synchronous_io, duplicate, &interrupted);
+	if (interrupted)
+		goto cleanup;
+
 	MONO_ENTER_GC_SAFE;
 	res = WriteFile (handle, buffer, numbytes, byteswritten, NULL);
 	MONO_EXIT_GC_SAFE;
+
+	mono_thread_info_uninstall_interrupt (&interrupted);
```

Comment:
The problem with `mono_thread_info_uninstall_interrupt` is that it races with `abort_w32_synchronous_io`, since the interrupt might happen at any moment. See https://github.com/mono/mono/blob/dde9ee5e79e756212625dc995ea5d2c0d479e372/mono/metadata/w32handle.c#L766 for how to protect yourself against that.

Reply:
Maybe the interrupt should take address of the local? And InterlockedExchange it? Might be easier to understand?

Reply:
What is the race? I am not closing the handle in the callback